File: duolingo_rum/Services/WordService.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using duolingo_rum.Models;
using Microsoft.EntityFrameworkCore;

namespace duolingo_rum.Services
{
    // ✅ ФИКС: WordService теперь IDisposable — контекст корректно освобождается
    public class WordService : IDisposable
    {
        private readonly _43pRumiantsefContext _context;
        private bool _disposed = false;

        public WordService()
        {
            _context = new _43pRumiantsefContext();
        }

        // ✅ НОВЫЙ МЕТОД: получаем свежие данные юзера из БД
        public async Task<User?> GetUserById(Guid userId)
        {
            try
            {
                // AsNoTracking чтобы не конфликтовало с уже трекаемыми объектами
                return await _context.Users
                    .AsNoTracking()
                    .FirstOrDefaultAsync(u => u.Id == userId);
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"GetUserById Error: {ex.Message}");
                return null;
            }
        }

        public async Task<List<Word>> GetWordsForLesson(Guid userId)
        {
            try
            {
                var words = await _context.Words
                    .Take(10)
                    .ToListAsync();

                return words ?? new List<Word>();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"GetWordsForLesson Error: {ex.Message}");
                return new List<Word>();
            }
        }

        public async Task<List<Word>> GetAnyWords(int count)
        {
            try
            {
                var words = await _context.Words
                    .Take(count)
                    .ToListAsync();

                return words ?? new List<Word>();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"GetAnyWords Error: {ex.Message}");
                return new List<Word>();
            }
        }

        public async Task SaveExerciseResult(Guid userId, int wordId, bool isCorrect, string userAnswer, string aiFeedback)
        {
            try
            {
                Debug.WriteLine($"SaveExerciseResult started for user {userId}, word {wordId}");

                var activeSession = await _context.LearningSessions
                    .FirstOrDefaultAsync(s => s.UserId == userId && s.FinishedAt == null);

                if (activeSession == null)
                {
                    activeSession = new LearningSession
                    {
                        Id = Guid.NewGuid(),
                        UserId = userId,
                        LanguageId = 2,
                        StartedAt = DateTime.UtcNow,
                        WordsStudied = 0,
                        CorrectAnswers = 0,
                        WrongAnswers = 0,
                        XpEarned = 0
                    };
                    _context.LearningSessions.Add(activeSession);
                    await _context.SaveChangesAsync();
                    Debug.WriteLine($"Created new session with ID: {activeSession.Id}");
                }

                var exerciseResult = new ExerciseResult
                {
                    SessionId = activeSession.Id,
                    WordId = wordId,
                    ExerciseType = "translation",
                    UserAnswer = userAnswer ?? string.Empty,
                    IsCorrect = isCorrect,
                    AiFeedback = aiFeedback ?? string.Empty,
                    AnsweredAt = DateTime.UtcNow,
                    ResponseTimeMs = 0
                };

                _context.ExerciseResults.Add(exerciseResult);
                Debug.WriteLine("Exercise result added to context");

                if (isCorrect)
                    activeSession.CorrectAnswers = (activeSession.CorrectAnswers ?? 0) + 1;
                else
                    activeSession.WrongAnswers = (activeSession.WrongAnswers ?? 0) + 1;

                activeSession.WordsStudied = (activeSession.WordsStudied ?? 0) + 1;
                activeSession.XpEarned = (activeSession.XpEarned ?? 0) + (isCorrect ? 10 : 5);

                await _context.SaveChangesAsync();
                Debug.WriteLine("Session stats saved");

                // ✅ ФИКС: используем Find (трекает по PK, не дублирует запросы)
                var user = await _context.Users.FindAsync(userId);
                if (user != null)
                {
                    user.TotalXp = (user.TotalXp ?? 0) + (isCorrect ? 10 : 5);
                    user.UpdatedAt = DateTime.UtcNow;
                    await _context.SaveChangesAsync();
                    Debug.WriteLine($"User XP updated to {user.TotalXp}");
                }

                try
                {
                    await UpdateWordProgressSM2(userId, wordId, isCorrect);
                }
                catch (Exception ex)
                {
                    Debug.WriteLine($"Error updating word progress: {ex.Message}");
                }

                Debug.WriteLine("SaveExerciseResult completed successfully");
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"SaveExerciseResult ERROR: {ex.Message}");
                if (ex.InnerException != null)
                    Debug.WriteLine($"Inner exception: {ex.InnerException.Message}");
                throw;
            }
        }

        private async Task UpdateWordProgressSimple(Guid userId, int wordId, bool isCorrect)
        {
            try
            {
                var progress = await _context.WordProgresses
                    .FirstOrDefaultAsync(wp => wp.UserId == userId && wp.WordId == wordId);

                if (progress == null)
                {
                    progress = new WordProgress
                    {
                        UserId = userId,
                        WordId = wordId,
                        Repetitions = 0,
                        EasinessFactor = 2.5m,
                        IntervalDays = 1,
                        CorrectCount = 0,
                        WrongCount = 0,
                        IsLearned = false,
                        CreatedAt = DateTime.UtcNow,
                        UpdatedAt = DateTime.UtcNow,
                        LastReview = DateOnly.FromDateTime(DateTime.UtcNow),
                        NextReview = DateOnly.FromDateTime(DateTime.UtcNow.AddDays(1))
                    };
                    _context.WordProgresses.Add(progress);
                    await _context.SaveChangesAsync();
                }

                if (isCorrect)
                {
                    progress.CorrectCount = (progress.CorrectCount ?? 0) + 1;
                    progress.Repetitions = (progress.Repetitions ?? 0) + 1;
                }
                else
                {
                    progress.WrongCount = (progress.WrongCount ?? 0) + 1;
                }

                progress.UpdatedAt = DateTime.UtcNow;
                await _context.SaveChangesAsync();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"UpdateWordProgressSimple Error: {ex.Message}");
            }
        }

        public async Task<int> GetLearnedWordsCount(Guid userId)
        {
            try
            {
                return await _context.WordProgresses
                    .Where(wp => wp.UserId == userId)
                    .CountAsync();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"GetLearnedWordsCount ERROR: {ex.Message}");
                return 0;
            }
        }

        public async Task<int> GetTodayProgress(Guid userId)
        {
            try
            {
                var today = DateTime.UtcNow.Date;

                var sessions = await _context.LearningSessions
                    .Where(s => s.UserId == userId)
                    .Select(s => s.Id)
                    .ToListAsync();

                return await _context.ExerciseResults
                    .Where(r => sessions.Contains(r.SessionId) &&
                                r.AnsweredAt.HasValue &&
                                r.AnsweredAt.Value.Date == today)
                    .CountAsync();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"GetTodayProgress ERROR: {ex.Message}");
                return 0;
            }
        }

        public async Task<List<Word>> GetWordsToReview(Guid userId, int count)
        {
            try
            {
                var words = await _context.WordProgresses
                    .Where(wp => wp.UserId == userId)
                    .Include(wp => wp.Word)
                    .Take(count)
                    .Select(wp => wp.Word)
                    .ToListAsync();

                return words ?? new List<Word>();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"GetWordsToReview ERROR: {ex.Message}");
                return new List<Word>();
            }
        }

        public async Task EndCurrentSession(Guid userId)
        {
            try
            {
                var activeSession = await _context.LearningSessions
                    .FirstOrDefaultAsync(s => s.UserId == userId && s.FinishedAt == null);

                if (activeSession != null)
                {
                    activeSession.FinishedAt = DateTime.UtcNow;
                    if (activeSession.StartedAt.HasValue)
                    {
                        activeSession.DurationSec = (int)(DateTime.UtcNow - activeSession.StartedAt.Value).TotalSeconds;
                    }
                    await _context.SaveChangesAsync();
                    Debug.WriteLine($"Session ended for user {userId}");
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"EndCurrentSession Error: {ex.Message}");
            }
        }

        // ✅ ФИКС: Dispose освобождает DbContext и закрывает соединение с БД
        public void Dispose()
        {
            if (!_disposed)
            {
                _context?.Dispose();
                _disposed = true;
            }
        }
        /// <summary>
        /// SRS: умный подбор слов — сначала те, у кого NextReview <= сегодня,
        /// потом новые (без записи в word_progress).
        /// </summary>
        public async Task<List<Word>> GetWordsForLessonSRS(Guid userId, int count = 10)
        {
            try
            {
                var today = DateOnly.FromDateTime(DateTime.UtcNow);

                // 1. Слова на повторение (просрочены или на сегодня)
                var reviewWords = await _context.WordProgresses
                    .Where(wp => wp.UserId == userId && wp.NextReview <= today)
                    .OrderBy(wp => wp.NextReview)
                    .Take(count)
                    .Include(wp => wp.Word)
                    .Select(wp => wp.Word)
                    .ToListAsync();

                if (reviewWords.Count >= count)
                    return reviewWords;

                // 2. Добираем новыми словами (которых нет в word_progress)
                var learnedWordIds = await _context.WordProgresses
                    .Where(wp => wp.UserId == userId)
                    .Select(wp => wp.WordId)
                    .ToListAsync();

                var newWords = await _context.Words
                    .Where(w => !learnedWordIds.Contains(w.Id))
                    .Take(count - reviewWords.Count)
                    .ToListAsync();

                reviewWords.AddRange(newWords);
                Debug.WriteLine($"SRS: {reviewWords.Count - newWords.Count} на повторение, {newWords.Count} новых");
                return reviewWords;
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"GetWordsForLessonSRS ERROR: {ex.Message}");
                return await GetWordsForLesson(userId); // фолбек на старый метод
            }
        }

        /// <summary>
        /// Обновляем интервал повторения по алгоритму SM-2.
        /// </summary>
        private async Task UpdateWordProgressSM2(Guid userId, int wordId, bool isCorrect)
        {
            try
            {
                var progress = await _context.WordProgresses
                    .FirstOrDefaultAsync(wp => wp.UserId == userId && wp.WordId == wordId);

                if (progress == null)
                {
                    progress = new WordProgress
                    {
                        UserId = userId,
                        WordId = wordId,
                        Repetitions = 0,
                        EasinessFactor = 2.5m,
                        IntervalDays = 1,
                        CorrectCount = 0,
                        WrongCount = 0,
                        IsLearned = false,
                        CreatedAt = DateTime.UtcNow,
                        UpdatedAt = DateTime.UtcNow,
                        LastReview = DateOnly.FromDateTime(DateTime.UtcNow),
                        NextReview = DateOnly.FromDateTime(DateTime.UtcNow.AddDays(1))
                    };
                    _context.WordProgresses.Add(progress);
                    await _context.SaveChangesAsync();
                }

                // SM-2 алгоритм
                if (isCorrect)
                {
                    progress.CorrectCount = (progress.CorrectCount ?? 0) + 1;
                    progress.Repetitions = (progress.Repetitions ?? 0) + 1;

                    int interval;
                    if (progress.Repetitions == 1) interval = 1;
                    else if (progress.Repetitions == 2) interval = 6;
                    else interval = (int)Math.Round((progress.IntervalDays ?? 1) * (double)(progress.EasinessFactor ?? 2.5m));

                    // Качество ответа: правильный = 4 (по шкале SM-2)
                    const double q = 4.0;
                    var ef = (double)(progress.EasinessFactor ?? 2.5m);
                    ef = ef + (0.1 - (5 - q) * (0.08 + (5 - q) * 0.02));
                    ef = Math.Max(1.3, ef);

                    progress.EasinessFactor = (decimal)Math.Round(ef, 2);
                    progress.IntervalDays = interval;
                    progress.IsLearned = progress.Repetitions >= 3;
                }
                else
                {
                    progress.WrongCount = (progress.WrongCount ?? 0) + 1;
                    // При ошибке сбрасываем — повторить завтра
                    progress.Repetitions = 0;
                    progress.IntervalDays = 1;
                }

                progress.LastReview = DateOnly.FromDateTime(DateTime.UtcNow);
                progress.NextReview = DateOnly.FromDateTime(DateTime.UtcNow.AddDays(progress.IntervalDays ?? 1));
                progress.UpdatedAt = DateTime.UtcNow;

                await _context.SaveChangesAsync();
                Debug.WriteLine($"SM-2 updated: interval={progress.IntervalDays}, ef={progress.EasinessFactor}, nextReview={progress.NextReview}");
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"UpdateWordProgressSM2 ERROR: {ex.Message}");
            }
        }

        /// <summary>
        /// Словарь: все слова пользователя с прогрессом, сгруппированные по топикам.
        /// </summary>
        public async Task<List<(Topic? Topic, List<(Word Word, WordProgress? Progress)> Words)>> GetVocabulary(Guid userId)
        {
            try
            {
                var allWords = await _context.Words
                    .Include(w => w.Topic)
                    .OrderBy(w => w.TopicId)
                    .ThenBy(w => w.Word1)
                    .ToListAsync();

                var progresses = await _context.WordProgresses
                    .Where(wp => wp.UserId == userId)
                    .ToListAsync();

                var progressMap = progresses.ToDictionary(p => p.WordId);

                var grouped = allWords
                    .GroupBy(w => w.Topic)
                    .Select(g => (
                        g.Key,
                        g.Select(w => (w, progressMap.TryGetValue(w.Id, out var p) ? p : null)).ToList()
                    ))
                    .ToList();

                return grouped;
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"GetVocabulary ERROR: {ex.Message}");
                return new();
            }
        }
    }
}

File: android-app/app/src/main/java/com/linguaai/app/services/WordService.kt
package com.linguaai.app.services

import android.util.Log
import com.linguaai.app.models.*
import io.github.jan.supabase.postgrest.postgrest
import java.time.Instant
import java.time.LocalDate
import java.util.UUID

class WordService {

    private val client = SupabaseClient.client
    private val tag = "WordService"

    suspend fun getUserById(userId: String): User? {
        return try {
            client.postgrest["users"]
                .select { filter { eq("id", userId) } }
                .decodeList<User>()
                .firstOrNull()
        } catch (e: Exception) {
            Log.e(tag, "getUserById Error: ${e.message}")
            null
        }
    }

    suspend fun getWordsForLesson(userId: String): List<Word> {
        return try {
            client.postgrest["words"]
                .select { limit(10) }
                .decodeList()
        } catch (e: Exception) {
            Log.e(tag, "getWordsForLesson Error: ${e.message}")
            emptyList()
        }
    }

    suspend fun getAnyWords(count: Int): List<Word> {
        return try {
            client.postgrest["words"]
                .select { limit(count.toLong()) }
                .decodeList()
        } catch (e: Exception) {
            Log.e(tag, "getAnyWords Error: ${e.message}")
            emptyList()
        }
    }

    suspend fun saveExerciseResult(
        userId: String,
        wordId: Int,
        isCorrect: Boolean,
        userAnswer: String,
        aiFeedback: String
    ) {
        try {
            var activeSession = client.postgrest["learning_sessions"]
                .select {
                    filter {
                        eq("user_id", userId)
                        isNull("finished_at")
                    }
                }
                .decodeList<LearningSession>()
                .firstOrNull()

            if (activeSession == null) {
                val newSession = LearningSession(
                    id = UUID.randomUUID().toString(),
                    userId = userId,
                    languageId = 2,
                    startedAt = Instant.now().toString(),
                    wordsStudied = 0,
                    correctAnswers = 0,
                    wrongAnswers = 0,
                    xpEarned = 0
                )
                client.postgrest["learning_sessions"].insert(newSession)
                activeSession = newSession
            }

            val exerciseResult = ExerciseResult(
                sessionId = activeSession.id,
                wordId = wordId,
                exerciseType = "translation",
                userAnswer = userAnswer,
                isCorrect = isCorrect,
                aiFeedback = aiFeedback,
                answeredAt = Instant.now().toString(),
                responseTimeMs = 0
            )
            client.postgrest["exercise_results"].insert(exerciseResult)

            val newCorrect = (activeSession.correctAnswers ?: 0) + if (isCorrect) 1 else 0
            val newWrong = (activeSession.wrongAnswers ?: 0) + if (!isCorrect) 1 else 0
            val newWordsStudied = (activeSession.wordsStudied ?: 0) + 1
            val xpGain = if (isCorrect) 10 else 5
            val newXp = (activeSession.xpEarned ?: 0) + xpGain

            client.postgrest["learning_sessions"].update(
                mapOf(
                    "correct_answers" to newCorrect,
                    "wrong_answers" to newWrong,
                    "words_studied" to newWordsStudied,
                    "xp_earned" to newXp
                )
            ) {
                filter { eq("id", activeSession.id) }
            }

            val user = getUserById(userId)
            if (user != null) {
                client.postgrest["users"].update(
                    mapOf(
                        "total_xp" to ((user.totalXp ?: 0) + xpGain),
                        "updated_at" to Instant.now().toString()
                    )
                ) {
                    filter { eq("id", userId) }
                }
            }

            updateWordProgress(userId, wordId, isCorrect)
        } catch (e: Exception) {
            Log.e(tag, "saveExerciseResult Error: ${e.message}")
        }
    }

    private suspend fun updateWordProgress(userId: String, wordId: Int, isCorrect: Boolean) {
        try {
            val existing = client.postgrest["word_progresses"]
                .select {
                    filter {
                        eq("user_id", userId)
                        eq("word_id", wordId)
                    }
                }
                .decodeList<WordProgress>()
                .firstOrNull()

            if (existing != null) {
                val newCorrect = (existing.correctCount ?: 0) + if (isCorrect) 1 else 0
                val newWrong = (existing.wrongCount ?: 0) + if (!isCorrect) 1 else 0
                val newReps = (existing.repetitions ?: 0) + 1
                val learned = newCorrect >= 3

                client.postgrest["word_progresses"].update(
                    mapOf(
                        "correct_count" to newCorrect,
                        "wrong_count" to newWrong,
                        "repetitions" to newReps,
                        "is_learned" to learned,
                        "last_review" to LocalDate.now().toString(),
                        "next_review" to LocalDate.now().plusDays(if (learned) 7 else 1).toString(),
                        "updated_at" to Instant.now().toString()
                    )
                ) {
                    filter { eq("id", existing.id) }
                }
            } else {
                val newProgress = WordProgress(
                    userId = userId,
                    wordId = wordId,
                    repetitions = 1,
                    correctCount = if (isCorrect) 1 else 0,
                    wrongCount = if (!isCorrect) 1 else 0,
                    isLearned = false,
                    lastReview = LocalDate.now().toString(),
                    nextReview = LocalDate.now().plusDays(1).toString(),
                    createdAt = Instant.now().toString(),
                    updatedAt = Instant.now().toString()
                )
                client.postgrest["word_progresses"].insert(newProgress)
            }
        } catch (e: Exception) {
            Log.e(tag, "updateWordProgress Error: ${e.message}")
        }
    }

    suspend fun getLearnedWordsCount(userId: String): Int {
        return try {
            client.postgrest["word_progresses"]
                .select {
                    filter {
                        eq("user_id", userId)
                    }
                }
                .decodeList<WordProgress>()
                .size
        } catch (e: Exception) {
            Log.e(tag, "getLearnedWordsCount Error: ${e.message}")
            0
        }
    }

    suspend fun getTodayProgress(userId: String): Int {
        return try {
            val today = LocalDate.now().toString()
            client.postgrest["word_progresses"]
                .select {
                    filter {
                        eq("user_id", userId)
                        eq("last_review", today)
                    }
                }
                .decodeList<WordProgress>()
                .size
        } catch (e: Exception) {
            0
        }
    }

    suspend fun getWordsToReview(userId: String, count: Int): List<Word> {
        return try {
            val today = LocalDate.now().toString()
            val progresses = client.postgrest["word_progresses"]
                .select {
                    filter {
                        eq("user_id", userId)
                        lte("next_review", today)
                    }
                    limit(count.toLong())
                }
                .decodeList<WordProgress>()

            if (progresses.isEmpty()) return emptyList()

            val wordIds = progresses.map { it.wordId }
            client.postgrest["words"]
                .select {
                    filter { isIn("id", wordIds) }
                }
                .decodeList()
        } catch (e: Exception) {
            Log.e(tag, "getWordsToReview Error: ${e.message}")
            emptyList()
        }
    }

    suspend fun finishSession(userId: String) {
        try {
            val session = client.postgrest["learning_sessions"]
                .select {
                    filter {
                        eq("user_id", userId)
                        isNull("finished_at")
                    }
                }
                .decodeList<LearningSession>()
                .firstOrNull() ?: return

            val startedAt = Instant.parse(session.startedAt ?: return)
            val duration = (Instant.now().epochSecond - startedAt.epochSecond).toInt()

            client.postgrest["learning_sessions"].update(
                mapOf(
                    "finished_at" to Instant.now().toString(),
                    "duration_sec" to duration
                )
            ) {
                filter { eq("id", session.id) }
            }
        } catch (e: Exception) {
            Log.e(tag, "finishSession Error: ${e.message}")
        }
    }

    data class VocabularyItem(val word: Word, val progress: WordProgress?)

    suspend fun getVocabulary(userId: String): List<Pair<Topic?, List<VocabularyItem>>> {
        return try {
            val words = client.postgrest["words"]
                .select()
                .decodeList<Word>()

            val progresses = client.postgrest["word_progresses"]
                .select { filter { eq("user_id", userId) } }
                .decodeList<WordProgress>()

            val topics = client.postgrest["topics"]
                .select()
                .decodeList<Topic>()

            val progressMap = progresses.associateBy { it.wordId }
            val topicMap = topics.associateBy { it.id }

            words.groupBy { it.topicId }
                .map { (topicId, wordList) ->
                    val topic = topicId?.let { topicMap[it] }
                    val items = wordList.map { w ->
                        VocabularyItem(w, progressMap[w.id])
                    }
                    topic to items
                }
        } catch (e: Exception) {
            Log.e(tag, "getVocabulary Error: ${e.message}")
            emptyList()
        }
    }
}
